### PR TITLE
Implement azure function endpoint to fetch all users

### PR DESCRIPTION
**Pull request by jak-ashuraliev**
issue #10

### CODE REVIEW

File: functions/HttpTriggerAPIUsers/__init__.py [hunk 1]
```diff
@@ -1,24 +1,52 @@
...
+    # Create a new connection
+    try:
+        with pypyodbc.connect(connection_string) as conn:
+            return select_users_query(conn)
+    except pypyodbc.DatabaseError as e:
+        if e.args[0] == '28000':
+          return func.HttpResponse(
+              "Unauthorized",
+              status_code=403
+          )
```
Comment (nataliagourova) [inline functions/HttpTriggerAPIUsers/__init__.py:31]: Nice error handling here

File: .gitignore [hunk 2]
```diff
@@ -102,3 +102,9 @@ dist
```
Comment (theKunte) [inline .gitignore:102]: I think the .vscode and the .venv code need to be added to the root .gitignore. I had some issue with that too

File: functions/HttpTriggerAPIUsers/__init__.py [hunk 3]
```diff
@@ -1,24 +1,52 @@
...
+        data = []
 
-    if name:
-        return func.HttpResponse(f"Hello, {name}. This HTTP triggered function executed successfully.")
-    else:
-        return func.HttpResponse(
-             "This HTTP triggered function executed successfully. Pass a name in the query string or in the request body for a personalized response.",
-             status_code=200
-        )
+        columns = [column[0] for column in cursor.description]
```
Comment (nataliagourova) [inline functions/HttpTriggerAPIUsers/__init__.py:47]: a comment on what the next few lines are doing would be helpful

File: functions/HttpTriggerAPIUsers/function.json [hunk 4]
```diff
@@ -9,7 +9,8 @@
       "methods": [
         "get",
         "post"
-      ]
+      ],
+      "route":"users"
```
Comment (nataliagourova) [inline functions/HttpTriggerAPIUsers/function.json:13]: good point! I will go ahead and add the route to my endpoint as well

**Review by nataliagourova — APPROVED** [review]

File: functions/HttpTriggerAPIUsers/__init__.py [hunk 5]
```diff
@@ -1,24 +1,52 @@
...
+    driver = '{ODBC Driver 17 for SQL Server}'
+    
+    # Define the connection string
+    connection_string = "Driver={};Server={};Database={};Uid={};Pwd={};Encrypt=yes;TrustServerCertificate=yes;Connection Timeout=30;".format(driver, server, database, username, password)
+    
+    # Create a new connection
+    try:
+        with pypyodbc.connect(connection_string) as conn:
+            return select_users_query(conn)
+    except pypyodbc.DatabaseError as e:
```
Comment (theKunte) [inline functions/HttpTriggerAPIUsers/__init__.py:26]: Nice! We might need to add additional error handling later  for the DB to get better responses for the user

File: functions/HttpTriggerAPIUsers/__init__.py [hunk 6]
```diff
@@ -1,24 +1,52 @@
...
-    name = req.params.get('name')
-    if not name:
-        try:
-            req_body = req.get_json()
-        except ValueError:
-            pass
-        else:
-            name = req_body.get('name')
+        # Get users
+        users = list(cursor.fetchall())
```
Comment (theKunte) [inline functions/HttpTriggerAPIUsers/__init__.py:39]: This fetches all user from the db! It is working!

File: functions/HttpTriggerAPIUsers/__init__.py (hunk 5 above)
Comment (jak-ashuraliev) [reply functions/HttpTriggerAPIUsers/__init__.py:26]: Thanks Jenny. I would add more error handling but I spent a lot of time trying to fix my issue with importing pyodbc and drivers.

File: functions/HttpTriggerAPIUsers/function.json (hunk 4 above)
Comment (jak-ashuraliev) [reply functions/HttpTriggerAPIUsers/function.json:13]: I believe Jenny, Ahmed and were able to rename the endpoints / routes.

File: .gitignore (hunk 2 above)
Comment (jak-ashuraliev) [reply .gitignore:102]: I have included these files in both places: inside and outside

File: functions/HttpTriggerAPIUsers/__init__.py (hunk 6 above)
Comment (jak-ashuraliev) [reply functions/HttpTriggerAPIUsers/__init__.py:39]: Tested with Farhad as well. It works.

**Review by farhadbahrehmandhenry — APPROVED** [review]

**Comment by jak-ashuraliev** [comment]
Sprint 1
Team 1
[Task] Implement Azure Function for /users API
Issue #10 

### Sprint 1 Goal is to implement HttpTrigger Azure Functions API for users

API endpoint: [/users](https://nsc-functionsapp-team1.azurewebsites.net/api/users) 

### Develop Azure Functions by using Visual Studio Code

Before you install and run the Azure Functions extension, you must meet these requirements:

- Visual Studio Code installed on one of the supported platforms.

- An active Azure subscription.

Please refer to the Microsoft documentation on how to install **Azure Function Extention** on Visual Studio Code.

### Prerequisites:
- Install Azure Functions Core Tools in Visual Studio Code
Reference: https://docs.microsoft.com/en-us/azure/azure-functions/functions-run-local?tabs=macos%2Ccsharp%2Cbash
- Install Azure Functions Extension in Visual Studio Code
Reference: https://docs.microsoft.com/en-us/azure/azure-functions/functions-develop-vs-code?tabs=csharp
- Install Python 
- Sign in to Azure 
- Verify and select the correct version of python 3.8 for your virtual env

Verify your ```local.settings.json``` contains the following:
  
"IsEncrypted": false,
  "Values": {
    "FUNCTIONS_WORKER_RUNTIME": "python",
    "ENV_DATABASE_USERNAME": "YOUR_DB_USERNAME",
    "ENV_DATABASE_PASSWORD": "YOUR_DB_PASSWORD",
    "ENV_DATABASE_NAME": "YOUR_DB_NAME",
    "ENV_DATABASE_SERVER": "YOUR_DB_SERVER_STRING"

### Time Spent

DATE | ACTIVITY | TIME
-- | -- | --
1/8 - 1/12 | Researched HttpTrigger Azure Functions | 4 hours
1/12 - 1/13 | Setup folder structure for the project and created functions | 1 hours
1/13 | Watched tutorials online | 2 hours
1/14 | Reseached wow to connect to Azure SQL Database  | 1 hour
1/15 |  Read documentation about PYODBC and how to install | 3 hour
1/17 - 1/19 | Blocker: issues with installing/importing pyodbc / ODBC Driver on macOS 11 | 20 hours
1/19 - 1/20 | Paid coding with team members | 4 hour
  | |Total Hours: 35

**Pull request merged**